taar/recommenders/similarity_recommender.py

```diff
@@ -0,0 +1,130 @@
+        return float(numer_val) / float(denum_val)
+
+    def compute_similarity_to_donors(self, client_data):
+        # Define which features have only categorical (unicode) value possibilities and which are continuous.
+        cat_feats = ["geoCity", "locale", "os"]
+        continuous_feats = ["subsessionLength", "bookmark_count", "tab_open_count", "total_uri", "unique_tlds"]
+        client_categorical_feats = [client_data.get(specified_key) for specified_key in cat_feats]
+        client_continuous_feats = [client_data.get(specified_key) for specified_key in continuous_feats]
+
+        s = []
+        for donor in self.donors_list:
+            # Separate out the categorical and numerical features for the donor.
+            donor_categorical_features = [donor.get(specified_key) for specified_key in cat_feats]
+            donor_continuous_features = [donor.get(specified_key) for specified_key in continuous_feats]
+
+            # Compute the similarity score between the donor and the client in the space of categorical features.
+            try:
+                # Here a larger distance indicates a poorer match between categorical variables.
+                j_d = (distance.hamming(client_categorical_feats, donor_categorical_features))
+            except ValueError:
+                logger.error("unable to compute similarity over categorical features for {}"
+                             .format(client_data["client_id"]))
+
+            # Compute the similarity score between the donor and the client in the space of numerical features.
+            try:
+                # Here a value close to zero indicates a good match in continuous variables.
+                j_c = (distance.canberra(client_continuous_feats, donor_continuous_features))
+            except ValueError:
+                logger.error("unable to compute similarity over continuous features for {}"
+                             .format(client_data["client_id"]))
+
+            # Take the product of similarities to attain a univariate similarity score.
+            # Addition of 0.001 to j_c avoids a zero value from the categorical variables, allowing j_d precedence
+            single_score = abs((j_c + 0.01) * j_d)
+
+            # compute the LR based on precomputed distributions that relate the score
+            #  to a probability of providing good addon recommendations
+            lr_from_score = self.get_lr(single_score)
+            s.append((lr_from_score, donor['activeAddons']))
+
+        return sorted(s, reverse=True)
+
+    def recommend(self, client_data, limit=10):
+        recommendations = []
+        s = self.compute_similarity_to_donors(client_data)
+        if s[0][0] < 1.5:
+            logger.warning("Addons recommended with very low similarity score, perhaps donor set is unrepresentative",
+                           extra={"max_sim": s[0]})
+
+        # Flatten output recommendations.
+        for addon_list in s:
```

Comment:
Just to make sure I understand: is this basically appending all the addons from all the donors, from the most similar one to the least similar one? If that's the case, maybe we should simply stop appending once we reach `limit`, instead of appending and then cropping the list before returning it.